src/util/http.go

```diff
@@ -0,0 +1,209 @@
+	post, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := NewRequest(http.MethodPost, address, bytes.NewReader(post))
+	if err != nil {
+		return nil, err
+	}
+
+	request.setContentType(mediatype.JSON)
+	request.setLength(len(post))
+
+	return request, nil
+}
+
+// XMLRequest returns a HTTPRequest using method POST with an XML marshalled body
+func XMLRequest(address string, body interface{}) (*HTTPRequest, error) {
+	// Strip the trailing newline, since we add our own later.
+	doctype := xml.Header[:len(xml.Header)-1]
+	return XMLRequestWithDoctype(address, doctype, body)
+}
+
+// XMLRequestWithDoctype returns a HTTPRequest using method POST with an XML marshalled body with the specified doctype
+func XMLRequestWithDoctype(address, doctype string, body interface{}) (*HTTPRequest, error) {
+	postBody, err := xml.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	post := bytes.Join([][]byte{[]byte(doctype), postBody}, []byte("\n"))
+
+	request, err := NewRequest(http.MethodPost, address, bytes.NewReader(post))
+	if err != nil {
+		return nil, err
+	}
+
+	request.setContentType(mediatype.XML)
+	request.setLength(len(post))
+
+	return request, nil
+}
+
+// FormRequest returns a HTTPRequest using method POST with a x-www-form-urlencoded marshalled body
+func FormRequest(address string, form map[string]string) (*HTTPRequest, error) {
+	post := urlValues(form).Encode()
+
+	request, err := NewRequest(http.MethodPost, address, strings.NewReader(post))
+	if err != nil {
+		return nil, err
+	}
+
+	request.setContentType(mediatype.Form)
+	request.setLength(len(post))
+
+	return request, nil
+}
+
+// URL returns the url.URL associated with this HTTPRequest
+func (r HTTPRequest) URL() *url.URL {
+	return r.Req.URL
+}
+
+// SetQuery sets a url query value on the HTTPRequest's URL
+func (r *HTTPRequest) SetQuery(key, value string) {
+	SetQuery(r.URL(), key, value)
+}
+
+// SetHeader sets a header on the HTTPRequest
+func (r *HTTPRequest) SetHeader(key, value string) {
+	r.Req.Header.Set(key, value)
+}
+
+func (r *HTTPRequest) setLength(length int) {
+	r.SetHeader("Content-Length", strconv.Itoa(length))
+}
+
+func (r *HTTPRequest) setContentType(mediaType mediatype.MediaType) {
+	r.SetHeader("Content-Type", mediaType.String())
+}
+
+// Accept sets the Accept header on the HTTPRequest to indicate what content-type we expect
+func (r *HTTPRequest) Accept(mediaType mediatype.MediaType) {
+	r.SetHeader("Accept", mediaType.String())
+}
+
+// Authorization sets the Authorization header on the HTTPRequest for token-based auth
+// e.g. request.Authorization("Bearer", token)
+func (r *HTTPRequest) Authorization(authType string, authKey string) {
+	r.SetHeader("Authorization", authType+" "+authKey)
+}
+
+// Do does a request and returns the response, as a HTTPResponse
+func (r *HTTPRequest) Do() (*HTTPResponse, error) {
+	resp, err := r.client.Do(r.Req)
+	return &HTTPResponse{resp}, err
```

Comment:
Should I be reading the response body to a `[]byte` here instead of using the Reader directly in `String()`, `JSON()` and `XML()`?